src/main/java/org/opensearch/ad/ratelimit/RateLimitedQueue.java

```diff
@@ -0,0 +1,465 @@
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.threadpool.ThreadPoolStats;
+
+/**
+ * HCAD can bombard Opensearch with “thundering herd” traffic, in which many entities
+ * make requests that need similar Opensearch reads/writes at approximately the same
+ * time. To remedy this issue we queue the requests and ensure that only a
+ * limited set of requests are out for Opensearch reads/writes.
+ *
+ * @param <RequestType> Individual request type that is a subtype of ADRequest
+ */
+public abstract class RateLimitedQueue<RequestType extends QueuedRequest> implements MaintenanceState {
+    /**
+     * Each request is associated with a segment. That is, a queue consists of segments.
+     * Segments have their corresponding priorities: HIGH, MEDIUM, and LOW. An example
+     * of HIGH priority requests is anomaly results with errors or its anomaly grade
+     * larger than zero. An example of MEDIUM priority requests is a cold start request
+     * for an entity. An example of LOW priority requests is checkpoint write requests
+     * for a cold entity. LOW priority requests have the slightest chance to be selected
+     * to be executed. MEDIUM and HIGH priority requests have higher stakes. LOW priority
+     * requests have higher chances of being deleted when the size of the queue reaches
+     * beyond a limit compared to MEDIUM/HIGH priority requests.
+     *
+     */
+    class Segment implements ExpiringState {
+        // last access time
+        private Instant lastAccessTime;
+        // data structure to hold requests. Cannot be reassigned.  This is to
+        // guarantee a segment's content cannot be null.
+        private final BlockingQueue<RequestType> content;
+
+        Segment() {
+            this.lastAccessTime = clock.instant();
+            this.content = new LinkedBlockingQueue<RequestType>();
+        }
+
+        @Override
+        public boolean expired(Duration stateTtl) {
+            return expired(lastAccessTime, stateTtl, clock.instant());
+        }
+
+        public void put(RequestType request) throws InterruptedException {
+            this.content.put(request);
+        }
+
+        public int size() {
+            return this.content.size();
+        }
+
+        /**
+         * This does not have to be precise, just a signal for unused old segment
+         * that can be removed.  It is fine if we have race condition.  Don't want
+         * to synchronize the access as this could penalize performance.
+         * @param lastAccessTime Last access time of the segment
+         */
+        public void setLastAccessTime(Instant lastAccessTime) {
+            this.lastAccessTime = lastAccessTime;
+        }
+    }
+
+    private static final Logger LOG = LogManager.getLogger(RateLimitedQueue.class);
+
+    protected int queueSize;
```

Comment:
Add `volatile` as it will be used in setting update consumer in line 159 ?

Reply:
added